tensorboard/backend/event_processing/data_provider.py

```diff
@@ -72,8 +72,69 @@ def list_runs(self, experiment_id):
     ]
 
   def list_scalars(self, experiment_id, plugin_name, run_tag_filter=None):
-    del experiment_id  # ignored for now
     run_tag_content = self._multiplexer.PluginRunToTagToContent(plugin_name)
+    return self._list(
+        provider.ScalarTimeSeries, run_tag_content, run_tag_filter
+    )
+
+  def read_scalars(
+      self, experiment_id, plugin_name, downsample=None, run_tag_filter=None
+  ):
+    # TODO(@wchargin): Downsampling not implemented, as the multiplexer
+    # is already downsampled. We could downsample on top of the existing
+    # sampling, which would be nice for testing.
+    del downsample  # ignored for now
+    index = self.list_scalars(
+        experiment_id, plugin_name, run_tag_filter=run_tag_filter
+    )
+
+    def convert_scalar_event(event):
+      return provider.ScalarDatum(
+          step=event.step,
+          wall_time=event.wall_time,
+          value=tensor_util.make_ndarray(event.tensor_proto).item(),
+      )
+
+    return self._read(convert_scalar_event, index)
+
+  def list_tensors(self, experiment_id, plugin_name, run_tag_filter=None):
```

Comment:
I generally don't think copy/paste is a bad thing but do we expect these implementations (list_tensors and read_tensors) to differ much beyond `TensorTimeSeries` and `convert_tensor_event`?

Reply:
I expect `list_tensors` and `read_tensors` to remain similar to their
scalar counterparts, though the scalar counterparts could gain
additional features, like “list scalar time series with tag name
`accuracy` whose moving average is above 0.999”, that don’t make sense
for general tensor time series. (This is why we chose to separate scalar
time series out in the first place, even though the basic functionality
is subsumed by tensor time series.)

In such a future, I’d be happy to inline `self._read` and `self._list`
with appropriate changes. If you’d prefer that these be inlined today,
that’s also fine with me. I’m not exactly sure what you’re asking, so
let me know if you’d like any changes?